processing/src/main/java/io/druid/segment/DimensionSelectors.java

```diff
@@ -0,0 +1,458 @@
+          new Function<String, DimensionSelector>()
+          {
+            @Override
+            public DimensionSelector apply(String dimension)
+            {
+              return DimensionSelectors.makeSelectorFromRow(
+                  rowSupplier,
+                  dimension,
+                  null
+              );
+            }
+          }
+      );
+
+      return DimensionSelectors.makeMultiDimensionalSelector(
+          selectors,
+          extractionFn
+      );
+    }
+  }
+
+  public static DimensionSelector makeSelectorFromRow(
+      final Supplier<Row> rowSupplier,
+      final String dimension,
+      final ExtractionFn extractionFn
+  )
+  {
+    return new DimensionSelector()
+    {
+      @Override
+      public IndexedInts getRow()
+      {
+        final List<String> dimensionValues = rowSupplier.get().getDimension(dimension);
+        final ArrayList<Integer> vals = Lists.newArrayList();
+        if (dimensionValues != null) {
+          for (int i = 0; i < dimensionValues.size(); ++i) {
+            vals.add(i);
+          }
+        }
+
+        return new IndexedInts()
+        {
+          @Override
+          public int size()
+          {
+            return vals.size();
+          }
+
+          @Override
+          public int get(int index)
+          {
+            return vals.get(index);
+          }
+
+          @Override
+          public Iterator<Integer> iterator()
+          {
+            return vals.iterator();
+          }
+
+          @Override
+          public void close() throws IOException
+          {
+
+          }
+
+          @Override
+          public void fill(int index, int[] toFill)
+          {
+            throw new UnsupportedOperationException("fill not supported");
+          }
+        };
+      }
+
+      @Override
+      public int getValueCardinality()
+      {
+        return DimensionSelector.CARDINALITY_UNKNOWN;
+      }
+
+      @Override
+      public String lookupName(int id)
+      {
+        final String value = rowSupplier.get().getDimension(dimension).get(id);
+        return extractionFn == null ? value : extractionFn.apply(value);
+      }
+
+      @Override
+      public int lookupId(String name)
+      {
+        if (extractionFn != null) {
+          throw new UnsupportedOperationException("cannot perform lookup when applying an extraction function");
+        }
+        return rowSupplier.get().getDimension(dimension).indexOf(name);
+      }
+    };
+  }
+
+  public static DimensionSelector makeMultiDimensionalSelector(
+      final List<DimensionSelector> selectors,
+      final ExtractionFn extractionFn
+  )
+  {
+    Preconditions.checkArgument(extractionFn != null && (extractionFn.arity() < 0 || extractionFn.arity() == selectors.size()),
+        "ExtractionFn should be specified and have the same arity as the number of dimensions");
+
+    return new DimensionSelector()
+    {
+      StringDimensionIndexer.DimensionDictionary dimLookup =
```

Comment:
This is going to add substantial heap pressure – it requires materializing a dictionary in heap for each multi-dimensional selector. I guess you're doing it this way because otherwise groupBy won't work properly, but we need to consider how we can do this better.